### PR TITLE
Configurable write logs to file

### PR DESCRIPTION
**Pull request by emostov**
# issue

The Sidecar team has been getting occasional reports of [`createType` issues](https://github.com/paritytech/substrate-api-sidecar/issues/253) from users. However, these errors are hard to reproduce and insufficient logging has been produced for them. An initial effort was made to create a simple CLI tool (see [here](https://github.com/danforbes/polkadot-js-scripts) & [here](https://github.com/emostov/polkadot-js-scripts) that would be used to capture the bare bones polkadot-js RPC calls that have been failing. The idea with logging RPC calls and responses is that we can look at the scale encoded values in response that failed on `createType` and check to see if the data is corrupt. Despite using the tools ourselves and some of Sidecars's users trying the tools on their setups we have not been able to log a `createType` error. With this lack of being able to reproduce the issue on demand, the importance of more robust, in production, logging functionality rises.

# presented solution

In this PR I introduce winston logging, which gives us rich configurability and a [wide host of transport options](https://github.com/winstonjs/winston/blob/master/docs/transports.md#winston-transports). Materially for the user, this modifies the console output to include time stamps and log levels, while also allowing logging to file. I introduce a host of configuration options for log levels and various specifics of file logging (see `README.md` additions for details).

All file logging is in a structured, JSON format that at minimum includes a timestamp, message, and level field.

In the process I have removed the morgan logger and replaced it with express-winston logger. The format for how requests are logged is now slightly changed in the console and and a large amount of info is now dumped to file for each request. I could not get the status code numbers color coded like before (not _important_ but a nice touch), however the log level is still color coded.

Our configuration manager only allows one to set their `.env` file with `NODE_ENV`, which is a name space collision with what we need for @polkadot-js/api RPC logging (`NODE_ENV=test`). The solution chevdor presented is to symlink your desired `.env` file with `.env.test`, I detail this process in the README

## logging RPC

The logging we care most about comes from @polkadot-js/api. In order to enable the logging one must set `NODE_ENV=test` and package will `console.log` too `stdout`.

 As far as I can tell we cannot put our own logging utilities into a dependency like @polkadot-js/api. So the only option I could think of was monkey patching `console.{log, error, warn}` with a function that wraps the equivalent winston logger level (see: src/logging/consoleOverride.ts). In the monkey path I had to change how winston formats inputs so it would behave like node.js's console.

The script `yarn start:log-rpc` was added to the `package.json` and it sets `NODE_ENV=test` when running `main.js`, so RPC logs will be output from @polkadot-js/api.

---

### other logging/config options we could consider:
- Look through options for file transport: https://github.com/winstonjs/winston/blob/master/docs/transports.md#file-transport
- Option to always write to same file (maybe helpful for those trying to redirect output?
- Rotating files option? David mentioned sys admins might want to do this themselves, but could perhaps could still be helpful for some?

### points for feedback
- Are logging options clear?
- Does the file logging setup allow afford our users the correct amount and type of control?
- Should we try to log any more or less info (I see the most place for change in what we log to file for HTTP requests)?

### CODE REVIEW

**Review by danforbes — APPROVED** [review]

**Comment by chevdor** [comment]
I think that writing the logs yourself is re-inventing the wheel heavily. See below.

Regarding the `.env` vs `.env.test` issue, my recommendation is to set `NODE_ENV` to what you need: `NODE_ENV=test` and if you want to use `.env` instead of `.env.test, you can create a symlink: `ln -s .env .env.test`.

To solve your logging problem in a clean way, I would suggest using a logger such as winston where you can log things without having to care about WHERE it lands and then tell the logger to write files, or write to a DB, to a log manager, etc....

If you need a sample, you can check out:
- definition: https://gitlab.com/Polkabot/polkabot/-/blob/master/packages/polkabot-api/src/LoggerFactory.ts
- instantiation: https://gitlab.com/Polkabot/polkabot/-/blob/master/packages/polkabot/src/index.ts#L17
- usage: https://gitlab.com/Polkabot/polkabot/-/blob/master/packages/polkabot/src/index.ts#L81

**Comment by dvdplm** [comment]
I second @chevdor here: I'd prefer an off-the-shelf logger. Not familiar with the node ecosystem though so I have no opinion of `winston`. Good loggers take care of nasty stuff like working with log rotation, crashes and platform dependent details and should be pretty easy to use.

**Comment by emostov** [comment]
@chevdor, @dvdplm - all good points that I agree with. The thing that stopped me from that originally using winston is that we need logging from a dependency and I can't put a third party logging statement directly in the dependencies code. But thinking about it I could wrap the logger in a function with the same signature as console.log and feed all the console statements through that - so I'll give that a try unless there is a better way.

File: specs.yml [hunk 1]
```diff
@@ -1,10 +1,39 @@
...
+  LOG:
+    CONSOLE_LEVEL:
+      description: Log level for console transport
+      default: info
+      regexp: ^error|warn|info|http|verbose|debug|silly$,
+
+    FILE_USE:
+      description: Wether or not to use file transport
+      default: yes
+      type: string
```
Comment (chevdor) [inline specs.yml:11]: This is probably where you should use a `boolean` instead of a string.
Cherry on top, `confmgr` will handle true/false as well as True/False, 0/1 and also yes/no.
Reference: https://gitlab.com/chevdor/confmgr/-/blob/master/src/ConfigManager.ts#L117

**Review by chevdor — CHANGES_REQUESTED** [review]
I suggest switching from `string` to `boolean`. See details below.

File: src/logging/consoleOverride.ts [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+		//
+		// `(format.apply(format, args as [string]) as unknown) as object: TS incorrectly says the we
+		// need a object as an argument to `info.call`. However, it will except a string perfectly fine,
+		// which is what `format.apply` returns.
+		logger.info.call(
+			logger,
+			(format.apply(format, args as [string]) as unknown) as object
+		);
+	};
+	console.info = function (...args: unknown[]) {
```
Comment (chevdor) [inline src/logging/consoleOverride.ts:27]: This function has a lot of duplication.
Basically, what you are doing is to say: "for function A, B and C, please use this implementation".

 You can make it shorter and more concise using:
```
['info', 'error', 'warn', 'log'].map (level => {
     console[level] = function (...);
}
```

**Comment by emostov** [comment]
See updated top comment that reflects recent changes.

Here is a snippet of what logging looks like in the console: 
Screen Shot 2020-09-15 at 12 28 42 AM

For now I chose to put a stack trace dump in the console whenever available since it can be helpful for tracking down the source of an issue, especially when users are asking for help debugging. It is less effort than asking someone to look in a file. However, I also think its reasonable to not include the stack trace by default in the console.

File: src/logging/transformers/stripAnsi.ts [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+}
+
+/**
+ * Strip ANSI characters from strings in simple, arbitrary data. N.B. this is not
+ * hardened to work with any possible javascript type and is only meant for arrays,
+ * basic objects and strings.
+ *
+ * @param data arbitrary data
+ */
+function sanitizeAnsi(data: unknown): unknown {
```
Comment (dvdplm) [inline src/logging/transformers/stripAnsi.ts:33]: I think `sanitize` is the wrong verb here: it's usually used to guard against security hazards. We're really stripping out data we don't want here, but not for security reasons. Can it be named `stripAnsiShellCodes` or some such?

It is called recursively (unbounded). I don't think that is a big concern, but maybe worth pondering for a second if users have any way to submit a malicious payload that would crash the server?

File: src/logging/transformers/stripAnsi.ts [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+const ansiRegex = new RegExp(pattern, 'g');
+
+/**
+ * Remove ANSI characters from a string by replacing them with empty an empty string.
+ *
+ * @param text
+ */
+function ansiReplace(text: unknown): typeof text {
+	return typeof text === 'string' ? text.replace(ansiRegex, '') : text;
+}
```
Comment (dvdplm) [inline src/logging/transformers/stripAnsi.ts:24]: Can't we assume `text` is a string here? The docs seem to do so already and the call site below also checks the type before calling. 

My TS is weak, but is the return value right here? Should it really be `typeof text`? Does that mean "the return type is whatever `text` was"?

File: src/logging/transformers/nodeUtilFormat.ts [hunk 5]
```diff
@@ -0,0 +1,19 @@
+import { TransformableInfo } from 'logform';
+import { SPLAT } from 'triple-beam';
+import util from 'util';
+import { format } from 'winston';
+
+/**
+ * Console.log style formatting using node's `util.format`. We need this so we
+ * can override console.{log, error, etc.} without issue.
+ */
+export const nodeUtilFormat = format(
```
Comment (dvdplm) [inline src/logging/transformers/nodeUtilFormat.ts:10]: I don't understand what this does, can you ELI5? What does `SPLAT` do?

File: src/logging/consoleOverride.ts [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+	[
+		['log', 'info'],
+		['info', 'info'],
+		['warn', 'warn'],
+		['error', 'error'],
+		['debug', 'debug'],
+	].forEach(([consoleLevel, winstonLevel]) => {
+		// Sacrereligious typecasting explained:
+		//
+		// `args as [string]`: fomat @types dictate that it needs an array of at least length 1. However,
```
Comment (dvdplm) [inline src/logging/consoleOverride.ts:22]:
```suggestion
		// `args as [string]`: format @types dictate that it needs an array of at least length 1. However,
```

File: src/logging/consoleOverride.ts [hunk 7]
```diff
@@ -0,0 +1,36 @@
...
+		// from testing this is not neccesary, so we override the type as a string tuple.
+		//
+		// `(format.apply(format, args as [string]) as unknown) as object: TS incorrectly says the we
```
Comment (dvdplm) [inline src/logging/consoleOverride.ts:25]:
```suggestion
		// `(format.apply(format, args as [string]) as unknown) as object`: TS incorrectly says the we
```

File: src/logging/consoleOverride.ts [hunk 8]
```diff
@@ -0,0 +1,36 @@
...
+		// need a object as an argument to `info.call`. However, it will except a string perfectly fine,
```
Comment (dvdplm) [inline src/logging/consoleOverride.ts:26]:
```suggestion
		// need a object as an argument to `info.call`. However, it will accept a string perfectly fine,
```

File: specs.yml [hunk 9]
```diff
@@ -1,10 +1,45 @@
...
+      type: number
+      regexp: "[0-9]+"
+
+    FILE_PATH:
+      description: Path to log file
+      default: ./logs/file-transport.log
+      type: string
+
+    FILE_STRIP_ANSI:
+      description: Wether or not to strip ANSI characters when writing to file transport
```
Comment (dvdplm) [inline specs.yml:38]:
```suggestion
      description: Whether or not to strip ANSI characters when writing to file transport
```

File: specs.yml [hunk 10]
```diff
@@ -1,10 +1,45 @@
...
-      regexp: ^error|all|none$
+  LOG:
+
+    CONSOLE_LEVEL:
+      description: Log level for console transport
+      default: info
+      regexp: ^error|warn|info|http|verbose|debug|silly$,
+
+    FILE_USE:
+      description: Wether or not to use file transport
```
Comment (dvdplm) [inline specs.yml:10]:
```suggestion
      description: Whether or not to use file transport
```

File: src/middleware/logger/httpLogger.ts [hunk 11]
```diff
@@ -0,0 +1,22 @@
...
+export const httpLoggerCreate = (winstonInstance: winston.Logger): e.Handler =>
+	expressWinston.logger({
+		// transports,
+		winstonInstance,
+		msg: `{{req.method}} {{req.url}} {{req.res.statusCode}} {{res.responseTime}}ms`,
+		level: (_req, res) => {
+			const { statusCode } = res;
+			if (statusCode < 400) {
+				// `http` is above `info`, so one needs to opt into `http` or above log level to view sub 400
+				return 'http';
```
[the remark this answers is not in the document]
Comment (emostov) [reply src/middleware/logger/httpLogger.ts:14]: See npm log levels: https://www.npmjs.com/package/winston#logging-levels

I don't thinks its particularly meaningful but it is the default

File: src/logging/transports/fileTransport.ts [hunk 12]
```diff
@@ -0,0 +1,37 @@
...
+export function fileTransport() {
+	const transformers = [
+		stripTimestamp(),
+		nodeUtilFormat(),
+		format.timestamp({ format: 'YYYY-MM-DD HH:mm:ss.SSS' }),
+		format.prettyPrint(),
+	];
+
+	if (config.FILE_STRIP_ANSI) {
+		console.info('STRIP');
```
[the remark this answers is not in the document]
Comment (emostov) [reply src/logging/transports/fileTransport.ts]: forgot to push my last clean up commit 🤦

File: src/logging/transports/fileTransport.ts [hunk 13]
```diff
@@ -0,0 +1,37 @@
...
+		transformers.unshift(stripAnsi());
+	} else {
+		console.debug('NO STRIP');
+	}
+
+	return new transports.File({
+		level: config.FILE_LEVEL || 'http',
+		filename: config.FILE_PATH || `./logs/file-transport.log`,
+		maxsize: config.FILE_SIZE || 524_288_000, // 500MB
+		maxFiles: config.FILE_COUNT || 2, // Max 1gb
```
[the remark this answers is not in the document]
Comment (emostov) [reply src/logging/transports/fileTransport.ts:29]: It does not stop logging, it will just delete the oldest file. Should we try to do something else?

File: src/logging/transports/consoleTransport.ts [hunk 14]
```diff
@@ -0,0 +1,38 @@
...
+export function consoleTransport() {
+	/**
+	 * A simple printing format for how `TransformableInfo` shows up.
+	 */
+	const simplePrint = format.printf((info: TransformableInfo) =>
+		info?.stack // If there is a stack dump (e.g. error middleware), show that in console
+			? `${info?.timestamp as string} ${info?.level}: ${
+					info?.message
+			  } \n ${info?.stack as string}`
+			: `${info?.timestamp as string} ${info?.level}: ${info?.message}`
```
[the remark this answers is not in the document]
Comment (emostov) [reply src/logging/transports/consoleTransport.ts:21]: Changed ... just wish we had Rust if else statements with implicit returns in TS

File: src/logging/transformers/stripAnsi.ts (hunk 4 above)
Comment (emostov) [reply src/logging/transformers/stripAnsi.ts]: You are correct in the use of `typeof`. 

I just removed this function since we where not reusing the logic anywhere else, and like you say we do a type check right before calling it.



I think when I initially added that function I thought I would use it in another way where it could be helpful to do runtime checks of the type to make sure it doesn't throw.

**Comment by emostov** [comment]
@dvdplm - I updated the config options for console to take into consideration those who use `stdout` to collect logging info (console transport goes to `stdout`). Changes reflected in the README.

Let me know when you think its ready to merge.

cc @joepetrowski

File: src/middleware/logger/httpLogger.ts (hunk 11 above)
[the remark this answers is not in the document]
Comment (dvdplm) [reply src/middleware/logger/httpLogger.ts:14]: TIL.

File: src/logging/transports/fileTransport.ts (hunk 13 above)
[the remark this answers is not in the document]
Comment (dvdplm) [reply src/logging/transports/fileTransport.ts:33]: It deletes the oldest and creates a new file? I'd prefer we don't try to manage log file size at all and just let `logrotate` (or whatever the node operator prefers) to do its thing. Having it here sure is convenient but I'm afraid it'll lead to issues down the line when `sidecar` tries to delete a file while some other process is busy reading/rotating it.

I think it's fine for now and no need to block this PR, but definitely something worth pointing out in the docs.

**Review by dvdplm — APPROVED** [review]

File: src/Config.ts [hunk 15]
```diff
@@ -8,9 +8,17 @@ import * as configTypes from '../config/types.json';
 export interface ISidecarConfig {
 	HOST: string;
 	PORT: number;
-	LOG_MODE: string;
 	WS_URL: string;
 	CUSTOM_TYPES: Record<string, string> | undefined;
+	CONSOLE_LEVEL: string;
```
Comment (chevdor) [inline src/Config.ts:13]: Regarding `confmgr`: your App is split into a few modules.
I think it could be helpful to keep this structure here instead of flattening it.
That also allows passing a whole 'LoggerConfig' Object to the logger related code.

So in short (pseudo code):
```
export SideCarConfigExpress {
  HOST: string;
  PORT: number;
}

export SideCarConfigLogs {
  ...
}


export interface ISidecarConfig {
    EXPRESS: SideCarConfigExpress;
    LOGS: SideCarConfigLogs;
    ....
}
```

File: src/logging/transports/fileTransport.ts (hunk 13 above)
[the remark this answers is not in the document]
Comment (emostov) [reply src/logging/transports/fileTransport.ts:33]: If we take out `maxSize` and `maxFile` then it will just keep writing to the same file. That would be a super quick change. We could also have yet another config option that just says "Only write to one file". Either way is ok for me. 
Honestly the main reason I had maxFile and maxSize is so some user who isn't really aware of the logs has random a random issue where Sidecar is sucking up disk space.

[the remark this answers is not in the document]
Comment (dvdplm) [reply src/logging/transports/fileTransport.ts:33]: > Honestly the main reason I had maxFile and maxSize is so some user who isn't really aware of the logs has random a random issue where Sidecar is sucking up disk space.

The default is to send logs to stdout only yes? If they opt in to log to file but do not read the docs and do not set up log rotation properly... well, that's on them.
